### ISSUE TITLE
How to Train aggressive overplaying bot geared towards beginner DDK levels?...

### ISSUE THREAD

**Issue by hydrogenpi** [gcp/leela-zero#1058]
What method would I go about if I wanted to create and train a new bot that is geared towards only beginner players, namely those in the ranks of 35 kyu to no higher than 10kyu.  This would be a strong bot that only played very weak human players (35 kyu to 10kyu) with the specific goal of winning by as many points as possible, and the ideal is to maybe even capture the entire board if capable. The intent and idea is for the bot to never relent and chase and try to kill everything on the board and never let off the gas against the weak human player. As a beginner we've all encountered the much stronger player that overplays aggressive moves, chases us all across the board and end up winning by huge bigly margins.. With all the proliferation of bots out there right now, no bot was trained or designed from ground up to play in such a manner! 

In the past I asked Ray developers how this would be possible and they stated that giving a high komi could emulate this but also it resulted in a early resignation and didn't exactly perform the way I expected. I believe to truly create a bot like this with this kind of aggressive behavior towards beginner/weak players is that it has to be trained that way from the start/ground up. 

From a high level can someone give me an idea what strategy I would use for to train and create a bot like this? 

=====

As for what is my motivation for considering doing something like this, well as a beginner player myself (around 21kyu OGS on a good day), I find that among the people that I interact with, 99% of the time the people who are exposed to this game only ever has a very casual interest and certainly almost none of the can play above 15kyu. In essence, for the vast majority of the cases, esp if Go becomes more popular and gain more exposure in the West and internationally globally, the common person isn't going to be able to appreciate the truth strength of a top end high level bot winning against a top pro vs the more immediate and Visceral effect of getting cleaned out and floor wiped by an overly aggressive bot, even if in terms of truth strength the aggressive bot is no where near as strong as a normal top level bot. The point is that I feel such a bot could deliver much more of a emotional WOW factor for beginner players in terms of the long hanging fruit than a ordinary stronger bot that play nuanced moves that such beginner player is not yet suited to appreciate that level yet. 

To give one example, a few years ago I played the first edition of CrazyStone DL v1.01 and set to 7D max strength. I was about 25kyu at the time. By playing very safe, I still lost the game, but the AI did't mess with me and I was able to "only" lose by 20 points or so. But this didn't represent the truth power of CSDL against me, because I know had it played a more aggressive style it would have gained far more points... and probably could have left me with nothing on the board.

**Comment by Yakago** [comment]
I guess one way would be to first create a bot ( or multiple bots) that play at the desired level.

Then take a 'self-play' approach where the new bot plays against the weaker bots and uses that data as training games. 

Perhaps you can use a dynamic komi approach, and gradually increase the komi as the network is able to beat the weaker bots.
The idea is that since the network is trained to beat the weak networks with high komi, it will be 'optimistic' in a real game and won't resign early.

**Comment by StanTraykov** [comment]
A fundamental difference seems to be that it can't self-train, as the bot would be too strong an opponent for itself--you'd need something weaker: another bot, a limited version of the same bot. And then it'd have to train for the highest score margin.

**Comment by hydrogenpi** [comment]
@StanTraykov Would it be better to do it as a 'zero' bot or supervised learning to start and then do self playing?
For example, what if I searched OGS/KGS for collection of all games, filtered by a weak player vs strong player where the total score was a huge margin of points won but the game lasted over at least say 100 moves... this sort of supervised learning for games tailored to this kind of criteria would initially create a bot that played with this style, and then I can continue by pitting a weaker version of this bot against a stronger version of itself to fine tune the process and make sure as the stronger version gets better that its goal is for max points etc.

**Comment by StanTraykov** [comment]
Hmm, wouldn't training a network for victory margin against itself at fewer playouts result all sorts of strange/unwanted things (such as playing to lose when behind?).

Since the original idea is to dish out justice at DDK players, maybe a static 5k-rated bot (e.g. po-limited LZ network that plays at 5k level--though if there are other options that closer to "human-like" 5k, they might be better) can be taken, and a network trained from zero for highest victory margin?

**Comment by Splee99** [comment]
Just an idea. We have so many historical training weights, basically from zero. Why do you need to train another one? Just pick any old weights trained before.

**Comment by Cabu** [comment]
We could probably train a new network against a bunch of weak bots where it get +1 (win) when it wipe the board and -1 (lose) if there is still stones from the weaker bots on the board. That way, you will probably have a network that would eleminate the other player from the board by playing in the weak point of the weak bots. By using enough different weak bots, we can probably emulate a standard weak player...

**Comment by hydrogenpi** [comment]
@StanTraykov good point, it might train the weaker bot to be even more weaker instead of the stronger bot to adapt a personality of overplay. Instead of starting from zero, a supervised learning bot from subset of all players/games where the much stronger player won by huge margins would at least give it that sort of personality to start out with...to make sure it goes down the right path...  then you are correct it has to play against a static bot, but maybe instead of a weaker bot this static bot can also be supervised learned from the subset of  all players/games where the much weaker player lost by huge margins.....

**Comment by lightvector** [comment]
Just throwing this out there - it's possible that you would get a reasonable bot for this purpose simply by using the AlphaGo-Zero process with no changes except to have the value objective be (+1 if win, -1 if loss) + C * (score difference) for a nonzero C. Along with a slight change to the value head so that it isn't constrained to return values in the range +1 to -1.

Past experience has shown that when using rollouts, anything more than a very small C hurts playing strength, and the same is probably true for a neural net. But I think you would get a bot that probably is still pretty strong. It would not "deliberately" overplay, but it also probably wouldn't slack off when ahead.

You could also provide C as an input to the neural net too and have the self-play games be played with a variety of values of C so the neural net gets experience with all different values. That way you get a bot that would be highly configurable on how much it cares about score maximizing.

**Comment by Dorus** [comment]
I think the easiest way to do this is to make a NN that predics policy and value for all different komi settings. During play, it will always use the policy net of the value net that gives over 50% winrate. How much over 50% depend on how difficult you want to make the bot, changing that would change the level of the bot. Going exactly for 50% encourages overplay, setting it conservative would make it play simpler moves when ahead like other MCTS bots do. Ofc it shouldnt accept komi values lower than it actually have, that would make it play to lose, but then i have a theory that doing so might actually result in a better game vs handicap stones.

The only problem will be to find enough training data. Self play is obvious an easy way to get data, however requires incredible amount of hardware, supervised training will be a lot easier to set up. If you use LeelaZ self play games you already have a lot of data, however all those games will be at around 7.5 komi, and they do not even have a final score as most end in resignation, so you would have to calculate an estimated score for them.

Especially since data is lacking on high or low komi games i do not expect this bot to become high dan level like LeelaZ, however against 10k players and lower like the goal is, it might do quite well.
